File: src/problemset/string/LongestCommonPrefix.go
package main

import (
	"fmt"
)

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	/* ans := ""
	for j := 0;; j++ {
		if j >= len(strs[0]) {
			break
		}
		pre := strs[0][j]
		for _, str := range strs {
			if j >= len(str) || str[j] != pre {
				return ans
			}
		}
		ans += string(pre)
	}
	return ans */
	for i := 0; i < len(strs[0]); i++ {
		for j := 1; j < len(strs); j++ {
			if i == len(strs[j]) || strs[j][i] != strs[0][i] {
				return strs[0][:i]
			}
		}
	}
	return strs[0]
}

File: src/problemset/string/LongestCommonPrefix.java
package problemset.string;

/**
 * @date: 2020-04-01
 * @time: 18:26:24
 * @author: MoeexT
 * @problem: 14
 * @url: https://leetcode-cn.com/problems/longest-common-prefix/
 */

public class LongestCommonPrefix {
    /** 
     * @time: 1ms 80.43%
     * @mem: 37.8MB 25.71%
     */
    public String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0) {
            return "";
        }
        StringBuilder sb = new StringBuilder();
        int col = Integer.MAX_VALUE;

        for (String string : strs) {
            if (string != null && string.length() != 0) {
                col = col < string.length() ? col : string.length();
            } else {
                return "";
            }
        }
        for (int i = 0; i < col; i++) {
            for (int j = 1; j < strs.length; j++) {
                // System.out.println(j + ", " + i);
                if (strs[j - 1].charAt(i) != strs[j].charAt(i)) {
                    return sb.toString();
                }
            }
            sb.append(strs[0].charAt(i));
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        LongestCommonPrefix solution = new LongestCommonPrefix();
        System.out.println(solution.longestCommonPrefix(new String[] { "aa", "a" }));
    }
}
